File: vlc-android/src/org/videolan/vlc/gui/browser/ExtensionAdapter.java
package org.videolan.vlc.gui.browser;

import androidx.databinding.DataBindingUtil;
import android.net.Uri;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.videolan.medialibrary.media.MediaWrapper;
import org.videolan.vlc.R;
import org.videolan.vlc.databinding.ExtensionItemViewBinding;
import org.videolan.vlc.extensions.api.VLCExtensionItem;
import org.videolan.vlc.media.MediaUtils;

import java.util.ArrayList;
import java.util.List;

public class ExtensionAdapter extends RecyclerView.Adapter<ExtensionAdapter.ViewHolder> {
    private static final String TAG = "VLC/ExtensionAdapter";

    ExtensionBrowser mFragment;
    List<VLCExtensionItem> mItemsList = new ArrayList<>();

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnLongClickListener {

        public ExtensionItemViewBinding binding;

        public ViewHolder(ExtensionItemViewBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            binding.setHolder(this);
        }

        public void onClick(View v){
            final VLCExtensionItem item = getItem(getLayoutPosition());
            if (item.type == VLCExtensionItem.TYPE_DIRECTORY) {
                mFragment.browseItem(item);
            } else if (item.type == VLCExtensionItem.TYPE_AUDIO || item.type == VLCExtensionItem.TYPE_VIDEO){
                MediaWrapper mw = new MediaWrapper(Uri.parse(item.link));
                mw.setDisplayTitle(item.getTitle());
                mw.setDescription(item.getSubTitle());
                mw.setType(getTypeAccordingToItem(item.type));
                MediaUtils.INSTANCE.openMedia(v.getContext(), mw);
            }
        }

        public void onMoreClick(View v){
            openContextMenu();
        }

        @Override
        public boolean onLongClick(View v) {
            return openContextMenu();
        }

        private boolean openContextMenu() {
            if (mFragment == null)
                return false;
            mFragment.openContextMenu(getLayoutPosition());
            return true;
        }
    }

    public ExtensionAdapter(ExtensionBrowser fragment) {
        mFragment = fragment;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ViewHolder((ExtensionItemViewBinding) DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()),R.layout.extension_item_view, parent, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final VLCExtensionItem item = getItem(position);
        holder.binding.setItem(item);
        holder.binding.executePendingBindings();
    }

    private int getIconResId(VLCExtensionItem item) {
        switch (item.type){
            case VLCExtensionItem.TYPE_AUDIO:
                return R.drawable.ic_browser_audio_normal;
            case VLCExtensionItem.TYPE_DIRECTORY:
                return R.drawable.ic_menu_folder;
            case VLCExtensionItem.TYPE_VIDEO:
                return R.drawable.ic_browser_video_normal;
            case VLCExtensionItem.TYPE_SUBTITLE:
                return R.drawable.ic_browser_subtitle_normal;
            default:
                return R.drawable.ic_browser_unknown_normal;
        }
    }

    public VLCExtensionItem getItem(int position) {
        return mItemsList.get(position);
    }

    public List<VLCExtensionItem> getAll() {
        return mItemsList;
    }

    @Override
    public int getItemCount() {
        return mItemsList.size();
    }

    public void addAll(List<VLCExtensionItem> list) {
        mItemsList.clear();
        mItemsList.addAll(list);
        notifyDataSetChanged();
    }

    public void clear() {
        mItemsList.clear();
    }

    private int getTypeAccordingToItem(int type) {
        switch (type) {
            case VLCExtensionItem.TYPE_DIRECTORY:
                return MediaWrapper.TYPE_DIR;
            case VLCExtensionItem.TYPE_VIDEO:
                return MediaWrapper.TYPE_VIDEO;
            case VLCExtensionItem.TYPE_AUDIO:
                return MediaWrapper.TYPE_AUDIO;
            case VLCExtensionItem.TYPE_PLAYLIST:
                return MediaWrapper.TYPE_PLAYLIST;
            case VLCExtensionItem.TYPE_SUBTITLE:
                return MediaWrapper.TYPE_SUBTITLE;
            default:
                return MediaWrapper.TYPE_ALL;
        }
    }
}

File: application/vlc-android/src/org/videolan/vlc/gui/browser/ExtensionAdapter.kt
package org.videolan.vlc.gui.browser

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.net.toUri
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.ObsoleteCoroutinesApi
import org.videolan.medialibrary.MLServiceLocator
import org.videolan.medialibrary.interfaces.media.MediaWrapper
import org.videolan.vlc.R
import org.videolan.vlc.databinding.ExtensionItemViewBinding
import org.videolan.vlc.extensions.api.VLCExtensionItem
import org.videolan.vlc.media.MediaUtils
import java.util.*

@ObsoleteCoroutinesApi
@ExperimentalCoroutinesApi
class ExtensionAdapter(internal var fragment: ExtensionBrowser?) : RecyclerView.Adapter<ExtensionAdapter.ViewHolder>() {
    private var itemsList: MutableList<VLCExtensionItem> = ArrayList()

    val all: List<VLCExtensionItem>
        get() = itemsList

    inner class ViewHolder(var binding: ExtensionItemViewBinding) : RecyclerView.ViewHolder(binding.root), View.OnLongClickListener {

        init {
            binding.holder = this
        }

        fun onClick(v: View) {
            val item = getItem(layoutPosition)
            if (item.type == VLCExtensionItem.TYPE_DIRECTORY) {
                fragment!!.browseItem(item)
            } else if (item.type == VLCExtensionItem.TYPE_AUDIO || item.type == VLCExtensionItem.TYPE_VIDEO) {
                val mw = MLServiceLocator.getAbstractMediaWrapper(item.link.toUri())
                mw.setDisplayTitle(item.getTitle())
                mw.description = item.getSubTitle()
                mw.type = getTypeAccordingToItem(item.type)
                MediaUtils.openMedia(v.context, mw)
            }
        }

        fun onMoreClick(v: View) {
            openContextMenu()
        }

        override fun onLongClick(v: View): Boolean {
            return openContextMenu()
        }

        private fun openContextMenu(): Boolean {
            if (fragment == null)
                return false
            fragment!!.openContextMenu(layoutPosition)
            return true
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(parent.context), R.layout.extension_item_view, parent, false) as ExtensionItemViewBinding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)
        holder.binding.item = item
        holder.binding.executePendingBindings()
    }

    private fun getIconResId(item: VLCExtensionItem): Int {
        return when (item.type) {
            VLCExtensionItem.TYPE_AUDIO -> R.drawable.ic_browser_audio_normal
            VLCExtensionItem.TYPE_DIRECTORY -> R.drawable.ic_menu_folder
            VLCExtensionItem.TYPE_VIDEO -> R.drawable.ic_browser_video_normal
            VLCExtensionItem.TYPE_SUBTITLE -> R.drawable.ic_browser_subtitle_normal
            else -> R.drawable.ic_browser_unknown_normal
        }
    }

    fun getItem(position: Int): VLCExtensionItem {
        return itemsList[position]
    }

    override fun getItemCount(): Int {
        return itemsList.size
    }

    fun addAll(list: List<VLCExtensionItem>) {
        itemsList.clear()
        itemsList.addAll(list)
        notifyDataSetChanged()
    }

    fun clear() {
        itemsList.clear()
    }

    private fun getTypeAccordingToItem(type: Int): Int {
        return when (type) {
            VLCExtensionItem.TYPE_DIRECTORY -> MediaWrapper.TYPE_DIR
            VLCExtensionItem.TYPE_VIDEO -> MediaWrapper.TYPE_VIDEO
            VLCExtensionItem.TYPE_AUDIO -> MediaWrapper.TYPE_AUDIO
            VLCExtensionItem.TYPE_PLAYLIST -> MediaWrapper.TYPE_PLAYLIST
            VLCExtensionItem.TYPE_SUBTITLE -> MediaWrapper.TYPE_SUBTITLE
            else -> MediaWrapper.TYPE_ALL
        }
    }

    companion object {
        private val TAG = "VLC/ExtensionAdapter"
    }
}
